### ISSUE TITLE
Can't convert model with IF op to quantized TF-Lite model

### ISSUE THREAD

**Issue by ScholliYT** [tensorflow/tensorflow#54304]
## System information
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Google Colab Ubuntu 18.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (or github SHA if from source): 2.7.0

## Provide the text output from tflite_convert

<details>
  <summary>console output of tflite converter</summary>

  ```py
  ---------------------------------------------------------------------------
  
  RuntimeError                              Traceback (most recent call last)
  
  [<ipython-input-12-1e0173995c78>](https://localhost:8080/#) in <module>()
       15 ]
       16 
  ---> 17 tflite_model = converter.convert()
       18 
       19 # Save the model.
  
  12 frames
  
  [/usr/local/lib/python3.7/dist-packages/tensorflow/lite/python/lite.py](https://localhost:8080/#) in convert(self)
     1669         Invalid quantization parameters.
     1670     """
  -> 1671     return super(TFLiteConverterV2, self).convert()
     1672 
     1673 
  
  [/usr/local/lib/python3.7/dist-packages/tensorflow/lite/python/lite.py](https://localhost:8080/#) in wrapper(self, *args, **kwargs)
      773   def wrapper(self, *args, **kwargs):
      774     # pylint: disable=protected-access
  --> 775     return self._convert_and_export_metrics(convert_func, *args, **kwargs)
      776     # pylint: enable=protected-access
      777 
  
  [/usr/local/lib/python3.7/dist-packages/tensorflow/lite/python/lite.py](https://localhost:8080/#) in _convert_and_export_metrics(self, convert_func, *args, **kwargs)
      759     self._save_conversion_params_metric()
      760     start_time = time.process_time()
  --> 761     result = convert_func(self, *args, **kwargs)
      762     elapsed_time_ms = (time.process_time() - start_time) * 1000
      763     if result:
  
  [/usr/local/lib/python3.7/dist-packages/tensorflow/lite/python/lite.py](https://localhost:8080/#) in convert(self)
     1339 
     1340     return super(TFLiteFrozenGraphConverterV2,
  -> 1341                  self).convert(graph_def, input_tensors, output_tensors)
     1342 
     1343 
  
  [/usr/local/lib/python3.7/dist-packages/tensorflow/lite/python/lite.py](https://localhost:8080/#) in convert(self, graph_def, input_tensors, output_tensors)
      950 
      951     return self._optimize_tflite_model(
  --> 952         result, self._quant_mode, quant_io=self.experimental_new_quantizer)
      953 
      954 
  
  [/usr/local/lib/python3.7/dist-packages/tensorflow/lite/python/convert_phase.py](https://localhost:8080/#) in wrapper(*args, **kwargs)
      224       except Exception as error:
      225         report_error_message(str(error))
  --> 226         raise error from None  # Re-throws the exception.
      227 
      228     return wrapper
  
  [/usr/local/lib/python3.7/dist-packages/tensorflow/lite/python/convert_phase.py](https://localhost:8080/#) in wrapper(*args, **kwargs)
      214     def wrapper(*args, **kwargs):
      215       try:
  --> 216         return func(*args, **kwargs)
      217       except ConverterError as converter_error:
      218         if converter_error.errors:
  
  [/usr/local/lib/python3.7/dist-packages/tensorflow/lite/python/lite.py](https://localhost:8080/#) in _optimize_tflite_model(self, model, quant_mode, quant_io)
      720         q_allow_float = quant_mode.is_allow_float()
      721         model = self._quantize(
  --> 722             model, q_in_type, q_out_type, q_activations_type, q_allow_float)
      723 
      724       m_in_type = in_type if in_type else _dtypes.float32
  
  [/usr/local/lib/python3.7/dist-packages/tensorflow/lite/python/lite.py](https://localhost:8080/#) in _quantize(self, result, input_type, output_type, activations_type, allow_float)
      540           self._experimental_disable_per_channel,
      541           input_data_type=input_type,
  --> 542           output_data_type=output_type)
      543     else:
      544       return calibrate_quantize.calibrate_and_quantize(
  
  [/usr/local/lib/python3.7/dist-packages/tensorflow/lite/python/convert_phase.py](https://localhost:8080/#) in wrapper(*args, **kwargs)
      224       except Exception as error:
      225         report_error_message(str(error))
  --> 226         raise error from None  # Re-throws the exception.
      227 
      228     return wrapper
  
  [/usr/local/lib/python3.7/dist-packages/tensorflow/lite/python/convert_phase.py](https://localhost:8080/#) in wrapper(*args, **kwargs)
      214     def wrapper(*args, **kwargs):
      215       try:
  --> 216         return func(*args, **kwargs)
      217       except ConverterError as converter_error:
      218         if converter_error.errors:
  
  [/usr/local/lib/python3.7/dist-packages/tensorflow/lite/python/convert.py](https://localhost:8080/#) in mlir_quantize(input_data_str, disable_per_channel, fully_quantize, inference_type, input_data_type, output_data_type, enable_numeric_verify, enable_whole_model_verify, denylisted_ops, denylisted_nodes)
      240       convert_tensor_tf_type_to_tflite_type(output_data_type),
      241       enable_numeric_verify, enable_whole_model_verify, denylisted_ops,
  --> 242       denylisted_nodes)
      243 
      244 
  
  [/usr/local/lib/python3.7/dist-packages/tensorflow/lite/python/wrap_toco.py](https://localhost:8080/#) in wrapped_experimental_mlir_quantize(input_data_str, disable_per_channel, fully_quantize, inference_type, input_data_type, output_data_type, enable_numeric_verify, enable_whole_model_verify, denylisted_ops, denylisted_nodes)
       48       input_data_str, disable_per_channel, fully_quantize, inference_type,
       49       input_data_type, output_data_type, enable_numeric_verify,
  ---> 50       enable_whole_model_verify, denylisted_ops, denylisted_nodes)
       51 
       52 
  
  RuntimeError: Failed to quantize: <unknown>:0: error: loc("Identity"): 'tf.If' op result #0 must be tensor of tf.dtype values, but got 'tensor<1x!quant.uniform<i8:f32, 0.0039215688593685627:-128>>'
  <unknown>:0: note: loc("Identity"): see current operation: %8 = "tf.If"(%6) {else_branch = @cond_false_100, is_stateless = false, then_branch = @cond_true_90} : (tensor<*xi1>) -> tensor<1x!quant.uniform<i8:f32, 0.0039215688593685627:-128>>
  ```
</details>

## Standalone code to reproduce the issue
[Colab notebook](https://colab.research.google.com/drive/1R3bIfsVqB1z6CuqJKLRj0oe_iCzOTP5t?usp=sharing)  

## TLDR
It is currently not possible to get a full `int8` quantized model with the `tfl.if` (`tf.cond`) control flow operation in it. The TFLiteConverter fails to convert. Dynamic quantization is working, however not helpful in my case as I want to run the model with tensorflow lite micro which doesn't support hybrid models.

**Comment by karimnosseir** [comment]
Jaesung can you help triage

**Comment by abattery** [comment]
@teijeong could you take a look? In my memory, you had a simple workaround solution.

@ScholliYT you can try "converter.experimental_new_quantizer = False" for the time being.

**Comment by abattery** [comment]
Thanks for the quick confirmation. TFLite full integer quantization has some limitations in supporting the control flow enabled graphs currently.

**Comment by ScholliYT** [comment]
Thanks for investigating @abattery. I just verified that this version converts the model to TF Lite (see colab notebook). The model also executes with TFLM 😃

**Comment by ScholliYT** [comment]
@abattery As I have continued to explore the topic here, I have noticed that quantization is not yet fully applied. With the latest version, tensors with dtype int8 are first dequantized to float32 and then passed to the subgraph of the IF operation. The subgraph then quantizes the tensor back to int8 and executes the operations in the subgraph with int8. 



```python
fc1_kernel = tf.Variable(tf.keras.initializers.GlorotUniform()([1,10]), name="FC1 Kernel", dtype=tf.float32)
fc2_kernel = tf.Variable(tf.keras.initializers.GlorotUniform()([1,10]), name="FC2 Kernel", dtype=tf.float32)

@tf.function(input_signature=[tf.TensorSpec(shape=[1], dtype=tf.float32)])
def ifop_example(x_input):

    x_input_batched = tf.reshape(x_input, [1,1])
    if x_input > tf.constant([0.5], dtype=tf.float32):
      y = tf.matmul(x_input_batched, fc1_kernel)
    else:
      y = tf.matmul(x_input_batched, fc2_kernel)
    return y
```

```md
Subgraph#0 main(T#0) -> [T#11]
  Op#0 GREATER(T#0, T#5) -> [T#6]
  Op#1 RESHAPE(T#6, T#1) -> [T#7]
  Op#2 RESHAPE(T#0, T#2) -> [T#8]
  Op#3 DEQUANTIZE(T#8) -> [T#9]        # dequantizes T8 before passing to the subgraph
  Op#4 IF(T#7, T#3, T#9, T#4, Then: Subgraph#2, Else: Subgraph#1) -> [T#10]
  Op#5 QUANTIZE(T#10) -> [T#11]        # quantizes T10 from subgraph back to int8, which was already int8 in T1_7 or T2_7

Tensors of Subgraph#0
  T#0(x_input) shape:[1], type:INT8
  T#1(Reshape_1/shape_1) shape:[0], type:INT32
  T#2(Reshape/shape) shape:[2], type:INT32 RO 8 bytes
  T#3(cond/input_0) shape:[1, 10], type:FLOAT32 RO 40 bytes
  T#4(cond/input_2) shape:[1, 10], type:FLOAT32 RO 40 bytes
  T#5(Const) shape:[1], type:INT8 RO 1 bytes
  T#6(Greater) shape:[1], type:BOOL
  T#7(Reshape_1) shape:[], type:BOOL
  T#8(Reshape) shape:[1, 1], type:INT8
  T#9(Reshape1) shape:[1, 1], type:FLOAT32
  T#10(Identity1) shape:[1, 10], type:FLOAT32
  T#11(Identity) shape:[1, 10], type:INT8

Subgraph#1 cond_false_frozen_5180(T#1_0, T#1_1, T#1_2) -> [T#1_8]
  Op#0 QUANTIZE(T#1_1) -> [T#1_4]
  Op#1 QUANTIZE(T#1_2) -> [T#1_5]
  Op#2 RESHAPE(T#1_5, T#1_3) -> [T#1_6]
  Op#3 FULLY_CONNECTED(T#1_4, T#1_6, T#-1) -> [T#1_7]
  Op#4 DEQUANTIZE(T#1_7) -> [T#1_8]

Tensors of Subgraph#1
  T#1_0(arg0) shape:[1, 10], type:FLOAT32
  T#1_1(arg1) shape:[1, 1], type:FLOAT32
  T#1_2(arg2) shape:[1, 10], type:FLOAT32
  T#1_3(cond/MatMul1) shape:[2], type:INT32 RO 8 bytes
  T#1_4(tfl.quantize) shape:[1, 1], type:INT8
  T#1_5(tfl.quantize1) shape:[1, 10], type:INT8
  T#1_6(cond/MatMul2) shape:[10, 1], type:INT8
  T#1_7(cond/MatMul3) shape:[1, 10], type:INT8
  T#1_8(tfl.dequantize) shape:[1, 10], type:FLOAT32

Subgraph#2 cond_true_frozen_5170(T#2_0, T#2_1, T#2_2) -> [T#2_8]
  Op#0 QUANTIZE(T#2_0) -> [T#2_4]
  Op#1 QUANTIZE(T#2_1) -> [T#2_5]
  Op#2 RESHAPE(T#2_4, T#2_3) -> [T#2_6]
  Op#3 FULLY_CONNECTED(T#2_5, T#2_6, T#-1) -> [T#2_7]
  Op#4 DEQUANTIZE(T#2_7) -> [T#2_8]

Tensors of Subgraph#2
  T#2_0(arg0) shape:[1, 10], type:FLOAT32
  T#2_1(arg1) shape:[1, 1], type:FLOAT32
  T#2_2(arg2) shape:[1, 10], type:FLOAT32
  T#2_3(cond/MatMul5) shape:[2], type:INT32 RO 8 bytes
  T#2_4(tfl.quantize2) shape:[1, 10], type:INT8
  T#2_5(tfl.quantize3) shape:[1, 1], type:INT8
  T#2_6(cond/MatMul6) shape:[10, 1], type:INT8
  T#2_7(cond/MatMul7) shape:[1, 10], type:INT8
  T#2_8(tfl.dequantize1) shape:[1, 10], type:FLOAT32
```

Any Idea how we could get full `int8` quantization? Or is there a good reason to convert back and forth here?
